### PR TITLE
Implement RecursiveDeletePatterns and delete BUILD files in kube rules

### PR DESCRIPTION
**Pull request by sttts**
Fixes https://github.com/kubernetes/publishing-bot/issues/49
Fixes https://github.com/kubernetes/client-go/issues/189

- [x] test this with a full run against kube

### CODE REVIEW

File: artifacts/scripts/util.sh [hunk 1]
```diff
@@ -533,8 +538,13 @@ function commit-subject() {
...
     local commit_msg_tag="${1}"
     local subdirectory="${2}"
+    local recursive_delete_pattern="${3}"
     echo "Running git filter-branch ..."
-    git filter-branch -f --msg-filter 'awk 1 && echo && echo "'"${commit_msg_tag}"': ${GIT_COMMIT}"' --subdirectory-filter "${subdirectory}" -- ${3} ${4} >/dev/null
+    local index_filter=""
+    if [ -n "${recursive_delete_pattern}" ]; then
+        index_filter="git rm --cached --ignore-unmatch -r ${recursive_delete_pattern}"
+    fi
+    git filter-branch -f --index-filter "${index_filter}" --msg-filter 'awk 1 && echo && echo "'"${commit_msg_tag}"': ${GIT_COMMIT}"' --subdirectory-filter "${subdirectory}" -- ${4} ${5} >/dev/null
```
Comment (sttts) [inline artifacts/scripts/util.sh]: we will see how much this slows down the publishing run

**Comment by sttts** [comment]
Tested a full run successfully. Here is the client-go BUILD removal: https://github.com/sttts/client-go/commit/d685b9c9f6e502a793237d3b9e0cb39529483f5f

File: configs/kubernetes-rules-configmap.yaml [hunk 2]
```diff
@@ -6,6 +6,9 @@ data:
     skip-source-branches:
     - release-1.5
     - release-1.6
+    recursive-delete-patterns:
+    - BUILD
```
Comment (ixdy) [inline configs/kubernetes-rules-configmap.yaml]: might want to also catch `BUILD.bazel` files.

Comment (sttts) [reply configs/kubernetes-rules-configmap.yaml]: That's only at the kubernetes root, isn't it?

Comment (sttts) [reply configs/kubernetes-rules-configmap.yaml]: I mean: not in staging/src/k8s.io/*

Comment (ixdy) [reply configs/kubernetes-rules-configmap.yaml]: right, we generally use `BUILD` in the kubernetes repo, though apparently `BUILD.bazel` is now preferred. I vaguely intend to switch at some point, but if/when I do that I guess I can also update the bot at the same time.

Comment (sttts) [reply configs/kubernetes-rules-configmap.yaml]: Then let's add it. It does not hurt.

Comment (sttts) [reply configs/kubernetes-rules-configmap.yaml]: done

**Comment by sttts** [comment]
@ixdy @caesarxuchao this needs a lgtm to deploy it for 1.10.

File: artifacts/scripts/util.sh [hunk 3]
```diff
@@ -243,6 +248,19 @@ sync_repo() {
...
+    # remove existing recursive-delete-pattern files. After a first removal commit, the filter-branch command
+    # will filter them out from upstream commits.
+    if [ -n "${recursive_delete_pattern}" ]; then
+        local split_recursive_delete_pattern
+        read -r -a split_recursive_delete_pattern <<< "${recursive_delete_pattern}"
+        git rm -q --ignore-unmatch -r "${split_recursive_delete_pattern[@]}"
+        git add -u
+        if ! git-index-clean; then
+            echo "Deleting files recursively: ${recursive_delete_pattern}"
+            git commit -q -m "sync: initially remove files ${recursive_delete_pattern}"
```
Comment (mfojtik) [inline artifacts/scripts/util.sh]: can the commit list the files it deleted? (instead of pattern) (or list the files being deleted in log).

**Comment by mfojtik** [comment]
one comment, LGTM otherwise

File: artifacts/scripts/util.sh (hunk 3 above)
Comment (sttts) [reply artifacts/scripts/util.sh]: That's one file per directory, and for each commit. That's spam the logs a lot.

Comment (sttts) [reply artifacts/scripts/util.sh]: actually, here we can be verbose. It's just the initial commit. Done.

**Pull request merged**